OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs

```diff
@@ -0,0 +1,199 @@
+			RenderContent.CellEntryChanged += UpdateSpriteLayers;
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		{
+			var resources = w.WorldActor.TraitsImplementing<ResourceType>()
+				.ToDictionary(r => r.Info.ResourceType, r => r);
+
+			// Build the sprite layer dictionary for rendering resources
+			// All resources that have the same palette must also share a sheet and blend mode
+			foreach (var r in resources)
+			{
+				var layer = spriteLayers.GetOrAdd(r.Value.Palette, pal =>
+					{
+						var first = r.Value.Variants.First().Value.First();
+						return new TerrainSpriteLayer(w, wr, first.Sheet, first.BlendMode, pal, wr.World.Type != WorldType.Editor);
+					});
+
+				// Validate that sprites are compatible with this layer
+				var sheet = layer.Sheet;
+				if (r.Value.Variants.Any(kv => kv.Value.Any(s => s.Sheet != sheet)))
+					throw new InvalidDataException("Resource sprites span multiple sheets. Try loading their sequences earlier.");
+
+				var blendMode = layer.BlendMode;
+				if (r.Value.Variants.Any(kv => kv.Value.Any(s => s.BlendMode != blendMode)))
+					throw new InvalidDataException("Resource sprites specify different blend modes. "
+						+ "Try using different palettes for resource types that use different blend modes.");
+			}
+
+			foreach (var cell in w.Map.AllCells)
+			{
+				var type = ResourceLayer.GetResourceType(cell);
+				if (type != null && info.RenderTypes.Contains(type.Info.Type))
+				{
+					// Initialize the RenderContent with the initial map state
+					// because the shroud may not be enabled.
+					var cellContent = CreateResourceCell(type, cell);
+					RenderContent[cell] = cellContent;
+					UpdateRenderedSprite(cell);
+				}
+			}
+		}
+
+		void UpdateSpriteLayers(CPos cell)
+		{
+			var type = ResourceLayer.GetResourceType(cell);
+			if (type != null && !info.RenderTypes.Contains(type.Info.Type))
+				return;
+
+			var resource = RenderContent[cell];
+			foreach (var kv in spriteLayers)
+			{
+				// resource.Type is meaningless (and may be null) if resource.Sprite is null
+				if (resource.Sprite != null && type.Palette == kv.Key)
+					kv.Value.Update(cell, resource.Sprite);
+				else
+					kv.Value.Update(cell, null);
+			}
+		}
+
+		void IRenderOverlay.Render(WorldRenderer wr)
+		{
+			foreach (var kv in spriteLayers.Values)
+				kv.Draw(wr.Viewport);
+		}
+
+		void ITickRender.TickRender(WorldRenderer wr, Actor self)
+		{
+			var remove = new List<CPos>();
+			foreach (var c in dirty)
+			{
+				if (!self.World.FogObscures(c))
+				{
+					var t = ResourceLayer.GetResourceType(c);
+					if (t != null)
+					{
+						var cellContent = CreateResourceCell(t, c);
+						RenderContent[c] = cellContent;
+					}
+
+					UpdateRenderedSprite(c);
+					remove.Add(c);
+				}
+			}
+
+			foreach (var r in remove)
+				dirty.Remove(r);
+		}
+
+		protected virtual void UpdateRenderedSprite(CPos cell)
+		{
+			var density = ResourceLayer.GetResourceDensity(cell);
```

Comment:
Querying Density / type / max density requires multiple cell layer lookups, which is bad for performance.  Call `GetResourceContent` once, and then check the struct members directly.